src/core/Akka.Streams/Implementation/Sources.cs

```diff
@@ -761,6 +763,274 @@ public UnfoldResourceSourceAsync(Func<Task<TSource>> create, Func<TSource, Task<
         public override string ToString() => "UnfoldResourceSourceAsync";
     }
 
+    /// <summary>
+    /// INTERNAL API
+    /// </summary>
+    /// <typeparam name="TOut">TBD</typeparam>
+    /// <typeparam name="TSource">TBD</typeparam>
+    /// <typeparam name="TCreateState">The state passed to resource create function</typeparam>
+    [InternalApi]
+    public sealed class UnfoldResourceSourceValueTaskAsync<TOut, TCreateState, TSource> : GraphStage<SourceShape<TOut>>
+    {
+        #region Logic
+
+        private sealed class Logic : OutGraphStageLogic
+        {
+            private readonly UnfoldResourceSourceValueTaskAsync<TOut, TCreateState, TSource> _stage;
+            private readonly Lazy<Decider> _decider;
+            private Option<TSource> _state = Option<TSource>.None;
+
+            private ValueTask<Option<TOut>> _currentReadVt;
+            private readonly Action _valueTaskAwaiterOnCompleteAction;
+            public Logic(UnfoldResourceSourceValueTaskAsync<TOut, TCreateState, TSource> stage, Attributes inheritedAttributes)
+                : base(stage.Shape)
+            {
+                _stage = stage;
+                _decider = new Lazy<Decider>(() =>
+                {
+                    var strategy = inheritedAttributes.GetAttribute<ActorAttributes.SupervisionStrategy>(null);
+                    return strategy != null ? strategy.Decider : Deciders.StoppingDecider;
+                });
+                _valueTaskAwaiterOnCompleteAction = SelfReadCallback;
```

Comment:
This pattern MAY, in fact, be too dangerous without an intermediate buffer or just wrapping the whole thing in `GetAsyncCallback(SelfReadCallback)`... happy for some input here.